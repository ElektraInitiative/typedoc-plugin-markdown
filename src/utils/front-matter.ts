import * as Handlebars from 'handlebars';
import { PageEvent } from 'typedoc/dist/lib/output/events';

export interface FrontMatterVars {
  [key: string]: string | number | boolean;
}

/**
 * Prepends YAML block to a string
 * @param contents - the string to prepend
 * @param vars - object of required front matter variables
 */
export const prependYAML = (contents: string, vars: FrontMatterVars) => {
  return contents
    .replace(/^/, toYAML(vars) + '\n\n')
    .replace(/[\r\n]{3,}/g, '\n\n');
};

/**
 * Returns the page title as rendered in the document h1(# title)
 * @param page
 */
export const getPageTitle = (page: PageEvent) => {
  return Handlebars.helpers.reflectionTitle.call(page, false);
};

/**
 * Converts YAML object to a YAML string
 * @param vars
 */
const toYAML = (vars: FrontMatterVars) => {
  const yaml = `---
${Object.entries(vars)
  .map(
    ([key, value]) =>
      `${key}: ${
        typeof value === 'string' ? `"${escapeString(value)}"` : value
      }`,
  )
  .join('\n')}
---`;
  return yaml;
};

// prettier-ignore
const escapeString=(str: string) => str.replace(/"/g, '\\"');
