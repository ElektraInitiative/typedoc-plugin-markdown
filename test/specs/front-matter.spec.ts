import { PageEvent, ProjectReflection } from 'typedoc';

import { getPageTitle, prependYAML } from '../../src/utils/front-matter';

describe(`FrontMatter:`, () => {
  let project: ProjectReflection;

  beforeAll(async () => {
    project = await global.bootstrap(['frontmatter.ts']);
  });

  test(`should prependYAML to doc content`, () => {
    expect(
      prependYAML('# Doc title\n\nDoc content', {
        stringProp: '"Escaped" title',
        booleanProp: true,
        numberProp: 2,
      }),
    ).toMatchSnapshot();
  });

  test(`should get page page from reflection helper`, () => {
    const page = {
      project: { url: 'index.md' },
      model: { name: 'Someclass' },
    } as PageEvent;
    getPageTitle(page);
    expect(getPageTitle(page)).toEqual('Someclass');
  });
});
