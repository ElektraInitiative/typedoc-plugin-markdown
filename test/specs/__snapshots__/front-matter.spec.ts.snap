// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FrontMatter: should prependYAML to doc content 1`] = `
"---
stringProp: "\\"Escaped\\" title"
booleanProp: true
numberProp: 2
---

# Doc title

Doc content"
`;
